File: server/advisor.js
import os from "node:os";
import { execFile } from "node:child_process";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

const FAMILY_CONTEXT_DEFAULTS = [
  [/llama\s*3\.1|llama3\.1/i, 128000],
  [/llama\s*3\.2|llama3\.2/i, 128000],
  [/qwen\s*2\.5|qwen2\.5/i, 32768],
  [/mistral|mixtral/i, 32768],
  [/gemma/i, 8192],
  [/phi/i, 128000],
];

export async function getSystemProfile() {
  const totalRamGb = bytesToGb(os.totalmem());
  const platform = os.platform();
  const gpu = await detectGpu(platform);

  return {
    platform,
    arch: os.arch(),
    cpu: os.cpus()?.[0]?.model ?? "Unknown CPU",
    cpuCores: os.cpus()?.length ?? null,
    totalRamGb,
    usableRamGb: round(totalRamGb * (platform === "darwin" ? 0.72 : 0.65), 1),
    gpu,
    memoryMode: platform === "darwin" ? "unified" : gpu.vramGb ? "gpu" : "system",
  };
}

export function profileManuscript(text, fileName) {
  const clean = text.replace(/\s+/g, " ").trim();
  const words = clean ? clean.split(/\s+/).length : 0;
  const estimatedTokens = Math.ceil(words * 1.32);
  const sections = estimateSections(text);

  return {
    fileName,
    characters: text.length,
    words,
    estimatedTokens,
    sections,
    manuscriptSize: sizeLabel(estimatedTokens),
  };
}

export function enrichModel(rawModel, showInfo = {}) {
  const name = rawModel.name ?? rawModel.model ?? "unknown";
  const details = showInfo.details ?? rawModel.details ?? {};
  const modelInfo = showInfo.model_info ?? {};
  const parameterCount =
    parseParameterCount(details.parameter_size) ??
    parseNumericParameterCount(modelInfo["general.parameter_count"]) ??
    parseParamsFromName(name);
  const quantization =
    details.quantization_level ??
    modelInfo["general.quantization_version"] ??
    parseQuantFromName(name) ??
    "Q4";
  const estimatedMemoryGb = estimateMemoryGb(parameterCount, quantization);
  const contextWindow =
    Number(modelInfo["llama.context_length"]) ||
    Number(modelInfo["qwen2.context_length"]) ||
    Number(modelInfo["gemma3.context_length"]) ||
    parseContextFromParameters(showInfo.parameters) ||
    inferContextWindow(name);

  return {
    name,
    modifiedAt: rawModel.modified_at,
    sizeGb: rawModel.size ? round(rawModel.size / 1024 ** 3, 1) : null,
    parameterCountB: parameterCount,
    parameterLabel: parameterCount ? `${trimDecimal(parameterCount)}B` : "Unknown",
    quantization,
    estimatedMemoryGb,
    contextWindow,
    family: details.family ?? modelInfo["general.architecture"] ?? inferFamily(name),
  };
}

export function scoreModels(models, systemProfile, manuscriptProfile) {
  return models
    .map((model) => scoreModel(model, systemProfile, manuscriptProfile))
    .sort((a, b) => b.score - a.score);
}

function scoreModel(model, systemProfile, manuscriptProfile) {
  const memoryPoolGb = systemProfile.gpu.vramGb || systemProfile.usableRamGb;
  const memoryRatio = model.estimatedMemoryGb ? model.estimatedMemoryGb / memoryPoolGb : 0.9;
  const promptReserve = 1800;
  const reviewBudget = Math.max(1024, model.contextWindow - promptReserve);
  const manuscriptRatio = manuscriptProfile.estimatedTokens / reviewBudget;

  const hardwareFit =
    memoryRatio <= 0.6 ? 100 : memoryRatio <= 0.85 ? 74 : memoryRatio <= 1 ? 48 : memoryRatio <= 1.35 ? 25 : 0;
  const contextFit =
    manuscriptRatio <= 0.8 ? 100 : manuscriptRatio <= 3 ? 76 : manuscriptRatio <= 9 ? 48 : 22;
  const quality = model.parameterCountB ? Math.min(100, 32 + Math.log2(model.parameterCountB + 1) * 18) : 52;
  const speed = model.parameterCountB ? Math.max(20, 100 - model.parameterCountB * 2.2 - memoryRatio * 12) : 50;
  const score = Math.round(hardwareFit * 0.38 + contextFit * 0.25 + quality * 0.22 + speed * 0.15);

  const fit = fitLabel(memoryRatio);
  const handling = handlingLabel(manuscriptRatio);
  const speedLabel = speed >= 76 ? "Fast" : speed >= 50 ? "Moderate" : "Slow";
  const recommendation = recommendationLabel(score, fit, handling);

  return {
    ...model,
    score,
    fit,
    speed: speedLabel,
    manuscriptHandling: handling,
    recommendation,
    reasons: buildReasons(model, fit, handling, memoryPoolGb, manuscriptProfile.estimatedTokens),
  };
}

function buildReasons(model, fit, handling, memoryPoolGb, tokens) {
  const reasons = [];
  if (model.estimatedMemoryGb) {
    reasons.push(`${model.estimatedMemoryGb} GB estimated model memory against ${round(memoryPoolGb, 1)} GB available pool`);
  }
  reasons.push(`${tokens.toLocaleString()} estimated manuscript tokens vs ${model.contextWindow.toLocaleString()} context`);
  reasons.push(`${fit} hardware fit and ${handling.toLowerCase()} manuscript handling`);
  return reasons;
}

function fitLabel(memoryRatio) {
  if (memoryRatio <= 0.6) return "Comfortable";
  if (memoryRatio <= 0.85) return "Usable";
  if (memoryRatio <= 1) return "Tight";
  if (memoryRatio <= 1.35) return "Slow fallback";
  return "Not recommended";
}

function handlingLabel(manuscriptRatio) {
  if (manuscriptRatio <= 0.8) return "Direct review";
  if (manuscriptRatio <= 3) return "Light chunking";
  if (manuscriptRatio <= 9) return "Multi-pass chunking";
  return "Heavy chunking";
}

function recommendationLabel(score, fit, handling) {
  if (fit === "Not recommended") return "Avoid on this machine";
  if (score >= 82) return "Best overall";
  if (score >= 68) return handling === "Direct review" ? "Strong choice" : "Recommended with chunking";
  if (score >= 50) return "Usable, expect tradeoffs";
  return "Only use if needed";
}

function estimateMemoryGb(parameterCountB, quantization) {
  if (!parameterCountB) return null;
  const quant = String(quantization).toUpperCase();
  const bits = quant.includes("Q8") ? 8 : quant.includes("Q6") ? 6 : quant.includes("Q5") ? 5 : quant.includes("Q3") ? 3.5 : 4.5;
  return round(parameterCountB * (bits / 8) * 1.22 + 0.8, 1);
}

function parseParameterCount(value) {
  if (!value) return null;
  const match = String(value).match(/([\d.]+)\s*B/i);
  return match ? Number(match[1]) : null;
}

function parseNumericParameterCount(value) {
  const n = Number(value);
  return Number.isFinite(n) && n > 0 ? round(n / 1e9, 1) : null;
}

function parseParamsFromName(name) {
  const match = String(name).match(/(?:^|[-_:])(\d+(?:\.\d+)?)\s*b(?:[-_:]|$)/i);
  return match ? Number(match[1]) : null;
}

function parseQuantFromName(name) {
  const match = String(name).match(/q[2-8](?:_[a-z])?/i);
  return match ? match[0].toUpperCase() : null;
}

function parseContextFromParameters(parameters = "") {
  const match = String(parameters).match(/num_ctx\s+(\d+)/i);
  return match ? Number(match[1]) : null;
}

function inferContextWindow(name) {
  for (const [pattern, context] of FAMILY_CONTEXT_DEFAULTS) {
    if (pattern.test(name)) return context;
  }
  return 8192;
}

function inferFamily(name) {
  return String(name).split(/[/:_-]/)[0] || "unknown";
}

function estimateSections(text) {
  const matches = text.match(/(?:^|\n)\s*(abstract|introduction|methods?|results?|discussion|conclusion|references)\s*(?:\n|$)/gi);
  return matches ? [...new Set(matches.map((m) => m.trim().toLowerCase()))].length : null;
}

function sizeLabel(tokens) {
  if (tokens < 7000) return "Short";
  if (tokens < 18000) return "Typical article";
  if (tokens < 45000) return "Long article";
  return "Very long manuscript";
}

async function detectGpu(platform) {
  if (platform === "darwin") {
    return detectMacGpu();
  }

  const nvidia = await tryExec("nvidia-smi", ["--query-gpu=name,memory.total", "--format=csv,noheader,nounits"]);
  if (nvidia) {
    const [name, memory] = nvidia.split("\n")[0].split(",").map((part) => part.trim());
    return { name, vramGb: memory ? round(Number(memory) / 1024, 1) : null, source: "nvidia-smi" };
  }

  return { name: "No discrete GPU detected", vramGb: null, source: "system" };
}

async function detectMacGpu() {
  const output = await tryExec("system_profiler", ["SPDisplaysDataType"]);
  if (!output) {
    return { name: "Apple unified memory", vramGb: null, source: "system" };
  }

  const chip = output.match(/Chipset Model:\s*(.+)/)?.[1]?.trim();
  const vram = output.match(/VRAM.*:\s*(\d+)\s*GB/i)?.[1];
  return {
    name: chip || "Apple unified memory",
    vramGb: vram ? Number(vram) : null,
    source: "system_profiler",
  };
}

async function tryExec(command, args) {
  try {
    const { stdout } = await execFileAsync(command, args, { timeout: 2500 });
    return stdout.trim();
  } catch {
    return "";
  }
}

function bytesToGb(bytes) {
  return round(bytes / 1024 ** 3, 1);
}

function round(value, digits = 0) {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

function trimDecimal(value) {
  return Number.isInteger(value) ? String(value) : String(value).replace(/\.0$/, "");
}

File: review_it/advisor.py
import math
import os
import platform
import re
import subprocess

import psutil


FAMILY_CONTEXT_DEFAULTS = [
    (re.compile(r"llama\s*3\.1|llama3\.1", re.I), 128000),
    (re.compile(r"llama\s*3\.2|llama3\.2", re.I), 128000),
    (re.compile(r"qwen\s*2\.5|qwen2\.5", re.I), 32768),
    (re.compile(r"mistral|mixtral", re.I), 32768),
    (re.compile(r"gemma", re.I), 8192),
    (re.compile(r"phi", re.I), 128000),
]


def get_system_profile() -> dict:
    total_ram_gb = round(psutil.virtual_memory().total / 1024**3, 1)
    system = platform.system().lower()
    gpu = detect_gpu(system)
    return {
        "platform": "darwin" if system == "darwin" else system,
        "arch": platform.machine(),
        "cpu": platform.processor() or platform.machine() or "Unknown CPU",
        "cpuCores": os.cpu_count(),
        "totalRamGb": total_ram_gb,
        "usableRamGb": round(total_ram_gb * (0.72 if system == "darwin" else 0.65), 1),
        "gpu": gpu,
        "memoryMode": "unified" if system == "darwin" else ("gpu" if gpu.get("vramGb") else "system"),
    }


def profile_manuscript(text: str, filename: str) -> dict:
    clean = re.sub(r"\s+", " ", text).strip()
    words = len(clean.split()) if clean else 0
    estimated_tokens = math.ceil(words * 1.32)
    return {
        "fileName": filename,
        "characters": len(text),
        "words": words,
        "estimatedTokens": estimated_tokens,
        "sections": estimate_sections(text),
        "manuscriptSize": size_label(estimated_tokens),
    }


def enrich_model(raw_model: dict, show_info: dict | None = None) -> dict:
    show_info = show_info or {}
    name = raw_model.get("name") or raw_model.get("model") or "unknown"
    details = show_info.get("details") or raw_model.get("details") or {}
    model_info = show_info.get("model_info") or {}
    parameter_count = (
        parse_parameter_count(details.get("parameter_size"))
        or parse_numeric_parameter_count(model_info.get("general.parameter_count"))
        or parse_params_from_name(name)
    )
    quantization = (
        details.get("quantization_level")
        or model_info.get("general.quantization_version")
        or parse_quant_from_name(name)
        or "Q4"
    )
    context_window = (
        as_int(model_info.get("llama.context_length"))
        or as_int(model_info.get("qwen2.context_length"))
        or as_int(model_info.get("gemma3.context_length"))
        or parse_context_from_parameters(show_info.get("parameters", ""))
        or infer_context_window(name)
    )
    return {
        "name": name,
        "modifiedAt": raw_model.get("modified_at"),
        "sizeGb": round(raw_model["size"] / 1024**3, 1) if raw_model.get("size") else None,
        "parameterCountB": parameter_count,
        "parameterLabel": f"{trim_decimal(parameter_count)}B" if parameter_count else "Unknown",
        "quantization": quantization,
        "estimatedMemoryGb": estimate_memory_gb(parameter_count, quantization),
        "contextWindow": context_window,
        "family": details.get("family") or model_info.get("general.architecture") or infer_family(name),
    }


def score_models(models: list[dict], system_profile: dict, manuscript_profile: dict) -> list[dict]:
    return sorted(
        (score_model(model, system_profile, manuscript_profile) for model in models),
        key=lambda item: item["score"],
        reverse=True,
    )


def score_model(model: dict, system_profile: dict, manuscript_profile: dict) -> dict:
    memory_pool_gb = system_profile.get("gpu", {}).get("vramGb") or system_profile["usableRamGb"]
    memory_ratio = (model.get("estimatedMemoryGb") or (memory_pool_gb * 0.9)) / memory_pool_gb
    review_budget = max(1024, model["contextWindow"] - 1800)
    manuscript_ratio = manuscript_profile["estimatedTokens"] / review_budget

    hardware_fit = 100 if memory_ratio <= 0.6 else 74 if memory_ratio <= 0.85 else 48 if memory_ratio <= 1 else 25 if memory_ratio <= 1.35 else 0
    context_fit = 100 if manuscript_ratio <= 0.8 else 76 if manuscript_ratio <= 3 else 48 if manuscript_ratio <= 9 else 22
    quality = min(100, 32 + math.log2(model["parameterCountB"] + 1) * 18) if model.get("parameterCountB") else 52
    speed_value = max(20, 100 - model["parameterCountB"] * 2.2 - memory_ratio * 12) if model.get("parameterCountB") else 50
    score = round(hardware_fit * 0.38 + context_fit * 0.25 + quality * 0.22 + speed_value * 0.15)

    fit = fit_label(memory_ratio)
    handling = handling_label(manuscript_ratio)
    speed = "Fast" if speed_value >= 76 else "Moderate" if speed_value >= 50 else "Slow"

    scored = dict(model)
    scored.update(
        {
            "score": score,
            "fit": fit,
            "speed": speed,
            "manuscriptHandling": handling,
            "recommendation": recommendation_label(score, fit, handling),
            "reasons": build_reasons(model, fit, handling, memory_pool_gb, manuscript_profile["estimatedTokens"]),
        }
    )
    return scored


def build_reasons(model: dict, fit: str, handling: str, memory_pool_gb: float, tokens: int) -> list[str]:
    reasons = []
    if model.get("estimatedMemoryGb"):
        reasons.append(f"{model['estimatedMemoryGb']} GB estimated model memory against {round(memory_pool_gb, 1)} GB available pool")
    reasons.append(f"{tokens:,} estimated manuscript tokens vs {model['contextWindow']:,} context")
    reasons.append(f"{fit} hardware fit and {handling.lower()} manuscript handling")
    return reasons


def fit_label(memory_ratio: float) -> str:
    if memory_ratio <= 0.6:
        return "Comfortable"
    if memory_ratio <= 0.85:
        return "Usable"
    if memory_ratio <= 1:
        return "Tight"
    if memory_ratio <= 1.35:
        return "Slow fallback"
    return "Not recommended"


def handling_label(manuscript_ratio: float) -> str:
    if manuscript_ratio <= 0.8:
        return "Direct review"
    if manuscript_ratio <= 3:
        return "Light chunking"
    if manuscript_ratio <= 9:
        return "Multi-pass chunking"
    return "Heavy chunking"


def recommendation_label(score: int, fit: str, handling: str) -> str:
    if fit == "Not recommended":
        return "Avoid on this machine"
    if score >= 82:
        return "Best overall"
    if score >= 68:
        return "Strong choice" if handling == "Direct review" else "Recommended with chunking"
    if score >= 50:
        return "Usable, expect tradeoffs"
    return "Only use if needed"


def estimate_memory_gb(parameter_count_b: float | None, quantization: str) -> float | None:
    if not parameter_count_b:
        return None
    quant = str(quantization).upper()
    bits = 8 if "Q8" in quant else 6 if "Q6" in quant else 5 if "Q5" in quant else 3.5 if "Q3" in quant else 4.5
    return round(parameter_count_b * (bits / 8) * 1.22 + 0.8, 1)


def parse_parameter_count(value) -> float | None:
    if not value:
        return None
    match = re.search(r"([\d.]+)\s*B", str(value), re.I)
    return float(match.group(1)) if match else None


def parse_numeric_parameter_count(value) -> float | None:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return round(number / 1e9, 1) if number > 0 else None


def parse_params_from_name(name: str) -> float | None:
    match = re.search(r"(?:^|[-_:])(\d+(?:\.\d+)?)\s*b(?:[-_:]|$)", name, re.I)
    return float(match.group(1)) if match else None


def parse_quant_from_name(name: str) -> str | None:
    match = re.search(r"q[2-8](?:_[a-z])?", name, re.I)
    return match.group(0).upper() if match else None


def parse_context_from_parameters(parameters: str) -> int | None:
    match = re.search(r"num_ctx\s+(\d+)", parameters or "", re.I)
    return int(match.group(1)) if match else None


def infer_context_window(name: str) -> int:
    for pattern, context in FAMILY_CONTEXT_DEFAULTS:
        if pattern.search(name):
            return context
    return 8192


def infer_family(name: str) -> str:
    return re.split(r"[/:_-]", name)[0] or "unknown"


def estimate_sections(text: str) -> int | None:
    matches = re.findall(r"(?:^|\n)\s*(abstract|introduction|methods?|results?|discussion|conclusion|references)\s*(?:\n|$)", text, re.I)
    return len({match.strip().lower() for match in matches}) or None


def size_label(tokens: int) -> str:
    if tokens < 7000:
        return "Short"
    if tokens < 18000:
        return "Typical article"
    if tokens < 45000:
        return "Long article"
    return "Very long manuscript"


def detect_gpu(system: str) -> dict:
    if system == "darwin":
        return detect_mac_gpu()

    nvidia = try_exec(["nvidia-smi", "--query-gpu=name,memory.total", "--format=csv,noheader,nounits"])
    if nvidia:
        name, _, memory = nvidia.splitlines()[0].partition(",")
        return {"name": name.strip(), "vramGb": round(float(memory.strip()) / 1024, 1) if memory.strip() else None, "source": "nvidia-smi"}

    return {"name": "No discrete GPU detected", "vramGb": None, "source": "system"}


def detect_mac_gpu() -> dict:
    output = try_exec(["system_profiler", "SPDisplaysDataType"])
    if not output:
        return {"name": "Apple unified memory", "vramGb": None, "source": "system"}
    chip = re.search(r"Chipset Model:\s*(.+)", output)
    vram = re.search(r"VRAM.*:\s*(\d+)\s*GB", output, re.I)
    return {
        "name": chip.group(1).strip() if chip else "Apple unified memory",
        "vramGb": int(vram.group(1)) if vram else None,
        "source": "system_profiler",
    }


def try_exec(command: list[str]) -> str:
    try:
        return subprocess.run(command, check=False, capture_output=True, text=True, timeout=2.5).stdout.strip()
    except Exception:
        return ""


def as_int(value) -> int | None:
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def trim_decimal(value: float) -> str:
    return str(int(value)) if float(value).is_integer() else str(value).rstrip("0").rstrip(".")
